### PR TITLE
mumps: import from homebrew/science

### PR DESCRIPTION
**Pull request by jslee02**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

~~Depends on #22361 and #22362~~

### CODE REVIEW

**Comment by fxcoudert** [comment]
As I said in the previous PR, let's make this formula optionless.
And that way, it won't depend on parmetis and scotch anymore (whose PR are far from ready)

**Comment by jslee02** [comment]
> As I said in the previous PR, let's make this formula optionless.

I'm open to making this formula optionless. But what if an user want to use `mumps` with `mpi` enabled?

Also, if I want to change this PR as so, do I need to make this PR as a single comment?

**Comment by fxcoudert** [comment]
Homebrew's approach is "one size fits all": try to give the best user experience to most our users, without source builds (and thus without options where possible).

You can make changes for now, and squash them all into a single commit at the end.

File: Formula/mumps.rb [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    doc.install Dir["doc/*.pdf"]
+    pkgshare.install "examples"
+
+    prefix.install "Makefile.inc"  # For the record.
+    File.open(prefix/"make_args.txt", "w") do |f|
+      f.puts(make_args.join(" "))  # Record options passed to make.
+    end
+  end
+
+  def caveats
```
Comment (fxcoudert) [inline Formula/mumps.rb]: Remove caveats.

File: Formula/mumps.rb [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      when building software that depends on MUMPS.
+    EOS
+    s
+  end
+
+  test do
+    cp_r pkgshare/"examples", testpath
+    opts = ["-I#{opt_include}", "-L#{opt_lib}", "-lmumps_common", "-lpord"]
+    opts << "-L#{Formula["openblas"].opt_lib}" << "-lopenblas"
+    if Tab.for_name("mumps").with?("mpi")
```
Comment (fxcoudert) [inline Formula/mumps.rb]: mumps doesn't have mpi anymore

File: Formula/mumps.rb [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+      mpirun = "mpirun -np 2"
+      opts << "-lscalapack"
+    else
+      f90 = ENV["FC"]
+      cc = ENV["CC"]
+      mpirun = ""
+    end
+
+    cd testpath/"examples" do
+      system f90, "-o", "ssimpletest", "ssimpletest.F", "-lsmumps", *opts
```
Comment (fxcoudert) [inline Formula/mumps.rb]: Remplace `f90` with `"gfortran"`

File: Formula/mumps.rb [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      system "#{mpirun} ./ssimpletest < input_simpletest_real"
+      system f90, "-o", "dsimpletest", "dsimpletest.F", "-ldmumps", *opts
+      system "#{mpirun} ./dsimpletest < input_simpletest_real"
+      system f90, "-o", "csimpletest", "csimpletest.F", "-lcmumps", *opts
+      system "#{mpirun} ./csimpletest < input_simpletest_cmplx"
+      system f90, "-o", "zsimpletest", "zsimpletest.F", "-lzmumps", *opts
+      system "#{mpirun} ./zsimpletest < input_simpletest_cmplx"
+      system cc, "-c", "c_example.c", "-I#{opt_include}"
```
Comment (fxcoudert) [inline Formula/mumps.rb]: Replace `cc` with `ENV.cc`

File: Formula/mumps.rb [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+      mpirun = ""
+    end
+
+    cd testpath/"examples" do
+      system f90, "-o", "ssimpletest", "ssimpletest.F", "-lsmumps", *opts
+      system "#{mpirun} ./ssimpletest < input_simpletest_real"
+      system f90, "-o", "dsimpletest", "dsimpletest.F", "-ldmumps", *opts
+      system "#{mpirun} ./dsimpletest < input_simpletest_real"
+      system f90, "-o", "csimpletest", "csimpletest.F", "-lcmumps", *opts
+      system "#{mpirun} ./csimpletest < input_simpletest_cmplx"
```
Comment (fxcoudert) [inline Formula/mumps.rb]: Remove `#{mpirun}`

File: Formula/mumps.rb [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    makefile = "Makefile.G95.SEQ"
+    cp "Make.inc/" + makefile, "Makefile.inc"
+
+    make_args += ["SCOTCHDIR=#{Formula["scotch"].opt_prefix}",
+                  "ISCOTCH=-I#{Formula["scotch"].opt_include}"]
+
+    scotch_libs = "LSCOTCH=-L$(SCOTCHDIR) -lscotch -lscotcherr -lscotcherrexit"
+    scotch_libs += "-lscotchmetis"
+    make_args << scotch_libs
+    orderingsf << " -Dptscotch"
```
Comment (fxcoudert) [inline Formula/mumps.rb]: We now build without scotch

File: Formula/mumps.rb [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+    make_args += ["LMETISDIR=#{Formula["metis"].opt_lib}",
+                  "IMETIS=#{Formula["metis"].opt_include}",
+                  "LMETIS=-L#{Formula["metis"].opt_lib} -lmetis"]
+    orderingsf << " -Dmetis"
+
+    make_args << "ORDERINGSF=#{orderingsf}"
+
+    make_args += ["CC=#{ENV["CC"]} -fPIC",
```
Comment (fxcoudert) [inline Formula/mumps.rb]: `-fPIC` is not needed here, as far as I can tell. It was useful on Linux, but not on macOS

File: Formula/mumps.rb [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+                  "FC=gfortran -fPIC",
+                  "FL=gfortran -fPIC"]
+
+    make_args << "LIBBLAS=-L#{Formula["openblas"].opt_lib} -lopenblas"
+
+    ENV.deparallelize # Build fails in parallel on Mavericks.
```
Comment (fxcoudert) [inline Formula/mumps.rb]: Was this bug reported?

File: Formula/mumps.rb [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+    system "make", "alllib", *(shlibs_args + make_args)
+
+    lib.install Dir["lib/*"]
```
Comment (fxcoudert) [inline Formula/mumps.rb]: Is there not a `make install` target we can run?

File: Formula/mumps.rb (hunk 8 above)
Comment (jslee02) [reply Formula/mumps.rb]: I took a look at the history of `mump.rb` in `homebrew/science` and found https://github.com/Homebrew/homebrew-science/commit/0e8c1ca803ba078c3822e24e26cf3627c608a016, but couldn't get what was the issue behind this line.

File: Formula/mumps.rb (hunk 9 above)
Comment (jslee02) [reply Formula/mumps.rb]: According to the makefile (i.e., [`Makefile.G95.SEQ`](https://github.com/group-gu/mumps/blob/master/Make.inc/Makefile.G95.SEQ)), it seems not.

**Comment by fxcoudert** [comment]
The more I'm looking at this, the messier it is… I think the `Makefile.inc.generic.SEQ` configuration file would be a better start than `g95`. And we should avoid shared libraries, until upstream decides to support them.

**Comment by jslee02** [comment]
> I think the Makefile.inc.generic.SEQ configuration file would be a better start than g95. And we should avoid shared libraries, until upstream decides to support them.

That sounds good to me. However, I'm not sure I could make those changes successfully. It seems out of my scope. If we need those changes for `mumps` to be merged, then I would like to find someone else who can handle this and wants to take over, which I would be grateful.

**Comment by fxcoudert** [comment]
I am going to close this for now. We need to improve this formula a lot before it can be incorporated into homebrew core.

**Pull request closed without merging**